scheduler/test/cook/test/compute_cluster.clj

```diff
@@ -0,0 +1,956 @@
+                        :state :running
+                        :state-locked? true
+                        :template "template1"}} (get-in-mem-configs)))))
+    (testing "exceptions"
+      (let [uri "datomic:mem://test-compute-cluster-config"
+            conn (restore-fresh-database! uri)]
+        (reset! cluster-name->compute-cluster-atom {})
+        (reset! initialize-cluster-fn-invocations-atom [])
+        (deliver exit-code-syncer-state-promise nil)
+        (deliver scheduler-promise nil)
+        (is (= {} (get-db-config-ents (d/db conn))))
+        (is (= {} (get-in-mem-configs)))
+        (is (= {:error-message "clojure.lang.ExceptionInfo: fail {}"
+                :update-succeeded false}
+               (execute-update! nil {:goal-config {:name "fail" :a :a :template "template1"} :valid? true :changed? true :active? true}
+                                (get-in-mem-configs))))
+        (is (= [] @initialize-cluster-fn-invocations-atom))
+        (is (= {} (get-db-config-ents (d/db conn))))
+        (is (= {} (get-in-mem-configs)))
+
+        (reset! initialize-cluster-fn-invocations-atom [])
+        (is (= {:error-message "java.lang.NullPointerException"
+                :update-succeeded false}
+               (execute-update! nil {:goal-config {:name "fail" :a :a :template "template1"} :valid? true :changed? true :active? true} nil)))
+        (is (= [] @initialize-cluster-fn-invocations-atom)))))
+
+  (testing "normal update"
+    (let [uri "datomic:mem://test-compute-cluster-config"
+          conn (restore-fresh-database! uri)]
+      (reset! initialize-cluster-fn-invocations-atom [])
+      (deliver exit-code-syncer-state-promise nil)
+      (deliver scheduler-promise nil)
+      (is (= {} (get-db-config-ents (d/db conn))))
+      (is (= {} (get-in-mem-configs)))
+      (with-redefs [config/compute-cluster-templates
+                    (constantly {"template1" {:config {:a :bb :c :dd
+                                                       :dynamic-cluster-config? true}
+                                              :e :ff
+                                              :factory-fn 'cook.test.compute-cluster/cluster-factory-fn}})]
+        (is (= {:update-succeeded true}
+               (execute-update! conn
+                                {:goal-config {:a :a
+                                               :name "name"
+                                               :template "template1"
+                                               :base-path "base-path"
+                                               :ca-cert "ca-cert"
+                                               :state :running
+                                               :state-locked? true}
+                                 :valid? true
+                                 :changed? true
+                                 :active? true}
+                                (get-in-mem-configs)))))
+      (is (= {:base-path "base-path"
```

Comment:
There's a lot of almost-identical config dictionaries in this code. Can they be refactored to some extent out with only the relevant differences hilighted?

Reply:
maybe, but I like to have more raw data in tests and as little code as possible

Reply:
Fair. All I'll say is think about if we really need this level of copy&paste. Its hard to see what the differences between these tests even are; I have to eyeball diff 40 lines of code each. (Also, FWIW, the unit test output shows the difference/computed result, so on a failure you know the result of the computation.)